gobblin-data-management/src/main/java/gobblin/data/management/copy/hive/HiveDatasetFinder.java

```diff
@@ -134,20 +157,32 @@ public boolean apply(String table) {
 
   @Override
   public Iterator<HiveDataset> getDatasetsIterator() throws IOException {
-    return Iterators.transform(getTables().iterator(), new Function<DbAndTable, HiveDataset>() {
+
+    return new AbstractIterator<HiveDataset>() {
+      private Iterator<DbAndTable> tables = getTables().iterator();
+
       @Override
-      public HiveDataset apply(@Nullable DbAndTable dbAndTable) {
-        if (dbAndTable == null) {
-          return null;
-        }
-        try (AutoReturnableObject<IMetaStoreClient> client = HiveDatasetFinder.this.clientPool.getClient()) {
-          Table table = client.get().getTable(dbAndTable.getDb(), dbAndTable.getTable());
-          return createHiveDataset(table);
-        } catch (IOException | TException ioe) {
-          throw new RuntimeException(ioe);
+      protected HiveDataset computeNext() {
+        while (this.tables.hasNext()) {
+          DbAndTable dbAndTable = this.tables.next();
+          try (AutoReturnableObject<IMetaStoreClient> client = HiveDatasetFinder.this.clientPool.getClient()) {
+            Table table = client.get().getTable(dbAndTable.getDb(), dbAndTable.getTable());
+            if (HiveDatasetFinder.this.eventSubmitter.isPresent()) {
```

Comment:
nit: create a method to optionally submit events which checks if `eventSubmitter` is present. So you don't need to have this `if` blocks twice.